0264-promise_gas_ratio.md

```diff
@@ -0,0 +1,145 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This host function would need to be implemented in `nearcore` and parallel [`promise_batch_action_function_call`](https://github.com/near/nearcore/blob/7d15bbc996282c8ae8f15b8f49d110fc901b84d8/runtime/near-vm-logic/src/logic.rs#L1526). Most details of these functions will be consistent, except that there will be additional bookkeeping for keeping track of which functions specified a ratio for unused gas. This will not affect or replace any existing host functions, but this will likely require a slightly higher gas cost than the original `promise_batch_action_function_call` host function due to this additional overhead.
+
+This host function definition would look like this (as a Rust consumer):
+```rust
+    /// Appends `FunctionCall` action to the batch of actions for the given promise pointed by
+    /// `promise_idx`. This function allows not specifying a specific gas value and allowing the
+    /// runtime to assign remaining gas based on a ratio.
+    ///
+    /// # Gas
+    ///
+    /// Gas can be specified using a static amount, a ratio of remaining prepaid gas, or a mixture
+    /// of both. To omit a static gas amount, [`u64::MAX`] can be passed for the `gas` parameter.
+    /// To omit assigning remaining gas, [`u64::MAX`] can be passed as the `gas_ratio` parameter.
+    ///
+    /// The gas ratio parameter works as the following:
+    ///
+    /// All unused prepaid gas from the current function call is split among all function calls
+    /// which supply this gas ratio. The amount attached to each respective call depends on the
+    /// value of the ratio.
+    ///
+    /// For example, if 40 gas is leftover from the current method call and three functions specify
+    /// the ratios 1, 5, 2 then 5, 25, 10 gas will be added to each function call respectively,
+    /// using up all remaining available gas.
+    ///
+    /// # Errors
+    ///
+    /// <...Ommitted previous errors as they do not change>
+    /// - If the [`u64::MAX`] special value is passed for `gas` and `gas_ratio` parameters
+    ///
+    ///
+    /// [`u64::MAX`]: std::u64::MAX
+    pub fn promise_batch_action_function_call_ratio(
+        promise_index: u64,
+        method_name_len: u64,
+        method_name_ptr: u64,
+        arguments_len: u64,
+        arguments_ptr: u64,
+        amount_ptr: u64,
+        gas: u64,
+        gas_ratio: u64,
+    );
+
+```
+
+The only difference from the existing API is `gas_ratio` added as another parameter, as an unsigned 64-bit integer.
+
+As for calculations, the remaining gas at the end of the transaction can be floor divided by the sum of all the ratios tracked. Then, after getting this value, just attach that value multiplied by the ratio gas to each function call action. 
+
+For example, if there are three ratios, `a`, `b`, `c`:
+```
+v = remaining_gas / (a + b + c)
```

Comment:
Can you specify how it works in terms of integer division? Do we always loose remainder of remaining gas by modulo (a + b + c + ...)? I think it should be ok for reasonable values of ratios, just want to make sure

Reply:
I mention a few lines above that it's floor division, but let me make it clear here also. As for u128 for the sum, good point I didn't include that information previously